### ISSUE TITLE
Code snippet control indentation

### ISSUE THREAD

**Issue by thetrompf** [Microsoft/vscode#20112]
How can I  control unindentation in code snippets?

```json
{
    "Create migration helper execute statement": {
        "prefix": "mhe",
        "body": [
            "$helper->execute(<<<SQL",
            "\t$0",
            "\rSQL",
            ");",
        ]
    }
}
```

When invoking it in indented context:

```php
class Migration {
→   public function up(Helper $helper) {
→   →   mhe<tab complete snippet>
→   }
}
```

Expected behavior:

```php
class Migration {
→   public function up(Helper $helper) {
→   →   $helper->execute(<<<SQL
→   →   →   |<cursor>
SQL
→   →   );
→   }
}
```

Actual behavior:

```php
class Migration {
→   public function up(Helper $helper) {
→   →   $helper->execute(<<<SQL
→   →   →   |<cursor>
→   →   
→   →   SQL
→   →   );
→   }
}
```

This is invalid behavior when using heredoc.

I tried to insert `\r` <carriage return> to try removing all indentation,
but is is just translated to a newline and indentation according to snippet context.

- VSCode Version: 1.9.0
- OS Version: macOS Sierra

**Comment by jrieken** [comment]
Tricky, today snippets are always adjusted to the indent of the line it is inserted at. Think of the normal for-loop. We might need some special character to signal this behaviour.

**Comment by thetrompf** [comment]
I thought that was what I was doing with `\r`?

**Comment by jrieken** [comment]
That's just interpreted as newline, like the elements in the `body` array...

**Comment by thetrompf** [comment]
Yes I'm aware :) what I meant was, that `\r` could be that "special" character?

**Comment by dakotak** [comment]
To me it seems more intuitive to be able to use `\b` in snippets _(handle ascii backspaces in the json string)_.

It would also be handy to have a special sequence to remove n levels of indentation, clear line, and have a reference to the starting indent level the cursor was on when the snippet was inserted.

**Comment by jrieken** [comment]
The question with the unindent is how far to go? Should `\b` unindent  by one indentation-unit or to the start of the line? Or should another character be use for start of like, like the start of text ascii control character?

**Comment by thetrompf** [comment]
I think, if you want to go that way, we need two special characters, one for start of line, and one for unindent one. Both are useful in different contexts. In my particular example I need, start of line in order to support heredoc.

**Comment by jlcvp** [comment]
I've came to an issue where I need to unindent the snippet by 1 indentation-unit. e.g.: imagine a lot of _elseif_ or python's _elif_ statements where the indentation is PART OF THE LANGUAGE.

\b support for unindent by 1 unit would be really great

**Comment by thetrompf** [comment]
I think support for unindent a single indentation unit will be great too. Sadly it does not solve this particular issue, because I don't how I need to unindent, I simply need to unindent to the start of the line, which `\r` represents. To be honest I don't really care what the token should be.

**Comment by Monk3yDev** [comment]
Is there any update for current case?

**Comment by Hezkore** [comment]
I've put `else` in both `increaseIndentPattern` and  `decreaseIndentPattern`.
When I type `else`; the indentation works perfectly.
But when using a snippet or auto-complete; it fails.

So I'd be happy if the snippet/auto-complete issue is at least fixed.

Example:
![image](https://user-images.githubusercontent.com/1727541/60351305-c852fe00-99c5-11e9-9a6a-afa137bbb1bb.png)

**Comment by thetrompf** [comment]
@Hezkore I see your issue, but I don't think it's related to this. This is an issue about "calling" formatters when a snippet is "completed" all placeholders has been replaced or the snippet is "tabbed away from".

**Comment by Hezkore** [comment]
@thetrompf Well... my bug report was closed in reference to this issue.
See #76139

**Comment by Hezkore** [comment]
This issue is still very much relevant. And I'm still not sure why my issue was closed if it's not similar to this one.
@jrieken any update on my issue (#76139) or why the merge happened?

**Comment by kripper** [comment]
Maybe related to https://github.com/microsoft/vscode/issues/171439 ?

**Comment by thetrompf** [comment]
> Maybe related to https://github.com/microsoft/vscode/issues/171439 ?

Maybe. If the issue you're referring to is solved in the way you suggest, it probably/maybe fixes this too.

But I don't agree that there is an issue in 171439. IMO if you select some code and instruct the editor to indent it, the editor shouldn't suddenly not "obey" - it is easy enough to only select the code you want to indent.

In this issue, at least how I imagined the solution, I just want snippets to adhere to the `\r` (carriage return) character.

**Comment by kripper** [comment]
> But I don't agree that there is an issue in 171439. IMO if you select some code and instruct the editor to indent it, the editor shouldn't suddenly not "obey" - it is easy enough to only select the code you want to indent.

This should be better discussed on that issue, but the thing is, it's very common having to move and indent big chunks of code and it's really annoying having to select and indent the code per blocks manually or to accidentally introduce syntax errors because you oversaw some HEREDOC tags.
You will never want to indent a HEREDOC tag because it simply breaks the code.

Indenting code is a formatting feature, and a formatting feature shouldn't break your code.

**Comment by jacekkopecky** [comment]
I too would like to see something like `\r` (and maybe the proposed `\b` added to handle indentation. My current need is a snippet that, whatever level of indentation I type it on, puts its content right at the start of the line.

**Comment by thegraffix** [comment]
> I too would like to see something like `\r` (and maybe the proposed `\b` added to handle indentation. My current need is a snippet that, whatever level of indentation I type it on, puts its content right at the start of the line.

👍 I agree! There needs to be a way to detect indentation in snippets, e.g. `${TM_CURRENT_INDENT}`
